<a href="https://colab.research.google.com/github/todalavibra/AI_SPRINT_2025_SmartVision_Counter_PaliGemma.ipynb/blob/main/Copy_of_Untitled1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Load data
from sklearn.datasets import load_breast_cancer
import pandas as pd

# Load the breast cancer dataset
breast_cancer = load_breast_cancer()

# Convert to a pandas DataFrame
df = pd.DataFrame(data=breast_cancer.data, columns=breast_cancer.feature_names)
df['target'] = breast_cancer.target

# Preprocess data
from sklearn.preprocessing import StandardScaler

# Separate features (X) and target (y)
X = df.drop('target', axis=1)
y = df['target']

# Import and instantiate StandardScaler
scaler = StandardScaler()

# Fit and transform the feature data
X_scaled = scaler.fit_transform(X)

# Choose and train model
from sklearn.ensemble import RandomForestClassifier

# Instantiate the model
model = RandomForestClassifier(random_state=42)

# Train the model
model.fit(X_scaled, y)

# Evaluate model
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Make predictions on the training data
y_pred = model.predict(X_scaled)

# Calculate and print the accuracy score
accuracy = accuracy_score(y, y_pred)
print(f'Accuracy: {accuracy:.4f}')

# Generate and print a classification report
print('Classification Report:')
print(classification_report(y, y_pred))

# Generate and print a confusion matrix
print('Confusion Matrix:')
print(confusion_matrix(y, y_pred))

# Make predictions (optional)
# Select a few rows as "new" data
new_data_df = df.sample(5, random_state=42)

# Separate features (X_new) and target (y_new)
X_new = new_data_df.drop('target', axis=1)
y_new = new_data_df['target']

# Scale the new data using the fitted scaler
X_new_scaled = scaler.transform(X_new)

# Make predictions on the scaled new data
predictions = model.predict(X_new_scaled)

# Print the predictions and compare with actual values
print("Predictions:", predictions)
print("Actual values:", y_new.values)

In [ ]:
# Load data
from sklearn.datasets import load_breast_cancer
import pandas as pd

# Load the breast cancer dataset
breast_cancer = load_breast_cancer()

# Convert to a pandas DataFrame
df = pd.DataFrame(data=breast_cancer.data, columns=breast_cancer.feature_names)
df['target'] = breast_cancer.target

# Preprocess data
from sklearn.preprocessing import StandardScaler

# Separate features (X) and target (y)
X = df.drop('target', axis=1)
y = df['target']

# Import and instantiate StandardScaler
scaler = StandardScaler()

# Fit and transform the feature data
X_scaled = scaler.fit_transform(X)

# Choose and train model
from sklearn.ensemble import RandomForestClassifier

# Instantiate the model
model = RandomForestClassifier(random_state=42)

# Train the model
model.fit(X_scaled, y)

# Evaluate model
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Make predictions on the training data
y_pred = model.predict(X_scaled)

# Calculate and print the accuracy score
accuracy = accuracy_score(y, y_pred)
print(f'Accuracy: {accuracy:.4f}')

# Generate and print a classification report
print('Classification Report:')
print(classification_report(y, y_pred))

# Generate and print a confusion matrix
print('Confusion Matrix:')
print(confusion_matrix(y, y_pred))

# Make predictions (optional)
# Select a few rows as "new" data
new_data_df = df.sample(5, random_state=42)

# Separate features (X_new) and target (y_new)
X_new = new_data_df.drop('target', axis=1)
y_new = new_data_df['target']

# Scale the new data using the fitted scaler
X_new_scaled = scaler.transform(X_new)

# Make predictions on the scaled new data
predictions = model.predict(X_new_scaled)

# Print the predictions and compare with actual values
print("Predictions:", predictions)
print("Actual values:", y_new.values)

# Task
Build and train a machine learning model in Google Colab.

## Load data

### Subtask:
Load a dataset suitable for a machine learning task (e.g., classification, regression).


**Reasoning**:
Load a publicly available dataset into a pandas DataFrame and display its head and info to understand its structure.



In [ ]:
from sklearn.datasets import load_breast_cancer
import pandas as pd

# Load the breast cancer dataset
breast_cancer = load_breast_cancer()

# Convert to a pandas DataFrame
df = pd.DataFrame(data=breast_cancer.data, columns=breast_cancer.feature_names)
df['target'] = breast_cancer.target

# Display the first few rows
display(df.head())

# Display information about the DataFrame
display(df.info())

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  target  
0          0.4601                  0.11890       0  
1          0.2750                  0.08902       0  
2          0.3613                  0.08758       0  
3          0.6638                  0.17300       0  
4          0.2364                  0.07678       0  

[5 rows x 31 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   mean radius              569 non-null    float64
 1   mean texture             569 non-null    float64
 2   mean perimeter           569 non-null    float64
 3   mean area                569 non-null    float64
 4   mean smoothness          569 non-null    float64
 5   mean compactness         569 non-null    float64
 6   mean concavity           569 non-null    float64
 7   mean concave points      569 non-null    float64
 8   mean symmetry            569 non-null    float64
 9   mean fractal dimension   569 non-null    float64
 10  radius error             569 non-null    float64
 11  texture error            569 non-null    float64
 12  perimeter error          569 non-null    float64
 13  area error               569 non-null    float64
 14  smoothness error         5

None

## Preprocess data

### Subtask:
Clean and prepare the data for model training (e.g., handle missing values, scale features, encode categorical variables).


**Reasoning**:
Separate features and target, then scale the features using StandardScaler.



In [ ]:
# Separate features (X) and target (y)
X = df.drop('target', axis=1)
y = df['target']

# Import and instantiate StandardScaler
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

# Fit and transform the feature data
X_scaled = scaler.fit_transform(X)

# Display the first few rows of the scaled data
display(X_scaled[:5])

array([[ 1.09706398e+00, -2.07333501e+00,  1.26993369e+00,
         9.84374905e-01,  1.56846633e+00,  3.28351467e+00,
         2.65287398e+00,  2.53247522e+00,  2.21751501e+00,
         2.25574689e+00,  2.48973393e+00, -5.65265059e-01,
         2.83303087e+00,  2.48757756e+00, -2.14001647e-01,
         1.31686157e+00,  7.24026158e-01,  6.60819941e-01,
         1.14875667e+00,  9.07083081e-01,  1.88668963e+00,
        -1.35929347e+00,  2.30360062e+00,  2.00123749e+00,
         1.30768627e+00,  2.61666502e+00,  2.10952635e+00,
         2.29607613e+00,  2.75062224e+00,  1.93701461e+00],
       [ 1.82982061e+00, -3.53632408e-01,  1.68595471e+00,
         1.90870825e+00, -8.26962447e-01, -4.87071673e-01,
        -2.38458552e-02,  5.48144156e-01,  1.39236330e-03,
        -8.68652457e-01,  4.99254601e-01, -8.76243603e-01,
         2.63326966e-01,  7.42401948e-01, -6.05350847e-01,
        -6.92926270e-01, -4.40780058e-01,  2.60162067e-01,
        -8.05450380e-01, -9.94437403e-02,  1.80592744e+

## Choose and train model

### Subtask:
Select an appropriate machine learning model and train it on the preprocessed data.


**Reasoning**:
Import and instantiate a RandomForestClassifier model and train it using the scaled feature data `X_scaled` and the target variable `y`.



In [ ]:
from sklearn.ensemble import RandomForestClassifier

# Instantiate the model
model = RandomForestClassifier(random_state=42)

# Train the model
model.fit(X_scaled, y)

RandomForestClassifier(random_state=42)

## Evaluate model

### Subtask:
Assess the performance of the trained model using relevant metrics.


**Reasoning**:
Import the necessary metrics from sklearn.metrics for classification evaluation.



In [ ]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

**Reasoning**:
Make predictions on the training data and calculate and print the accuracy score, classification report, and confusion matrix to assess the model's performance.



In [ ]:
# Make predictions on the training data
y_pred = model.predict(X_scaled)

# Calculate and print the accuracy score
accuracy = accuracy_score(y, y_pred)
print(f'Accuracy: {accuracy:.4f}')

# Generate and print a classification report
print('Classification Report:')
print(classification_report(y, y_pred))

# Generate and print a confusion matrix
print('Confusion Matrix:')
print(confusion_matrix(y, y_pred))

Accuracy: 1.0000
Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00       212
           1       1.00      1.00      1.00       357

    accuracy                           1.00       569
   macro avg       1.00      1.00      1.00       569
weighted avg       1.00      1.00      1.00       569

Confusion Matrix:
[[212   0]
 [  0 357]]


## Make predictions (optional)

### Subtask:
Use the trained model to make predictions on new data.


**Reasoning**:
Select a few rows from the original DataFrame to represent new data, separate features and target, scale the features using the previously fitted scaler, and then use the trained model to make predictions.



In [ ]:
# Select a few rows as "new" data
new_data_df = df.sample(5, random_state=42)

# Separate features (X_new) and target (y_new)
X_new = new_data_df.drop('target', axis=1)
y_new = new_data_df['target']

# Scale the new data using the fitted scaler
X_new_scaled = scaler.transform(X_new)

# Make predictions on the scaled new data
predictions = model.predict(X_new_scaled)

# Print the predictions and compare with actual values
print("Predictions:", predictions)
print("Actual values:", y_new.values)

Predictions: [1 0 0 1 1]
Actual values: [1 0 0 1 1]


## Summary:

### Data Analysis Key Findings

*   The breast cancer dataset containing 569 entries and 31 features was successfully loaded and preprocessed.
*   Features were scaled using `StandardScaler` to prepare the data for model training.
*   A `RandomForestClassifier` model was trained on the scaled data.
*   The model achieved perfect accuracy (1.0000) on the training data, with precision, recall, and F1-score of 1.00 for both classes.
*   The confusion matrix showed no misclassifications on the training data.
*   Predictions made on a small sample of "new" data matched the actual values.

### Insights or Next Steps

*   The perfect performance on the training data suggests potential overfitting. The next step should be to split the data into training and testing sets and evaluate the model's performance on unseen data to get a realistic assessment of its generalization ability.
*   Consider hyperparameter tuning of the `RandomForestClassifier` or exploring other models to potentially improve performance and generalization on unseen data.


## Load data

### Subtask:
Load a dataset suitable for a machine learning task (e.g., classification, regression).

**Reasoning**:
Load a publicly available dataset into a pandas DataFrame and display its head and info to understand its structure.

In [ ]:
from sklearn.datasets import load_breast_cancer
import pandas as pd

# Load the breast cancer dataset
breast_cancer = load_breast_cancer()

# Convert to a pandas DataFrame
df = pd.DataFrame(data=breast_cancer.data, columns=breast_cancer.feature_names)
df['target'] = breast_cancer.target

# Display the first few rows
display(df.head())

# Display information about the DataFrame
display(df.info())

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  target  
0          0.4601                  0.11890       0  
1          0.2750                  0.08902       0  
2          0.3613                  0.08758       0  
3          0.6638                  0.17300       0  
4          0.2364                  0.07678       0  

[5 rows x 31 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   mean radius              569 non-null    float64
 1   mean texture             569 non-null    float64
 2   mean perimeter           569 non-null    float64
 3   mean area                569 non-null    float64
 4   mean smoothness          569 non-null    float64
 5   mean compactness         569 non-null    float64
 6   mean concavity           569 non-null    float64
 7   mean concave points      569 non-null    float64
 8   mean symmetry            569 non-null    float64
 9   mean fractal dimension   569 non-null    float64
 10  radius error             569 non-null    float64
 11  texture error            569 non-null    float64
 12  perimeter error          569 non-null    float64
 13  area error               569 non-null    float64
 14  smoothness error         5

None

## Preprocess data

### Subtask:
Clean and prepare the data for model training (e.g., handle missing values, scale features, encode categorical variables).

**Reasoning**:
Separate features and target, then scale the features using StandardScaler.

In [ ]:
# Separate features (X) and target (y)
X = df.drop('target', axis=1)
y = df['target']

# Import and instantiate StandardScaler
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

# Fit and transform the feature data
X_scaled = scaler.fit_transform(X)

# Display the first few rows of the scaled data
display(X_scaled[:5])

array([[ 1.09706398e+00, -2.07333501e+00,  1.26993369e+00,
         9.84374905e-01,  1.56846633e+00,  3.28351467e+00,
         2.65287398e+00,  2.53247522e+00,  2.21751501e+00,
         2.25574689e+00,  2.48973393e+00, -5.65265059e-01,
         2.83303087e+00,  2.48757756e+00, -2.14001647e-01,
         1.31686157e+00,  7.24026158e-01,  6.60819941e-01,
         1.14875667e+00,  9.07083081e-01,  1.88668963e+00,
        -1.35929347e+00,  2.30360062e+00,  2.00123749e+00,
         1.30768627e+00,  2.61666502e+00,  2.10952635e+00,
         2.29607613e+00,  2.75062224e+00,  1.93701461e+00],
       [ 1.82982061e+00, -3.53632408e-01,  1.68595471e+00,
         1.90870825e+00, -8.26962447e-01, -4.87071673e-01,
        -2.38458552e-02,  5.48144156e-01,  1.39236330e-03,
        -8.68652457e-01,  4.99254601e-01, -8.76243603e-01,
         2.63326966e-01,  7.42401948e-01, -6.05350847e-01,
        -6.92926270e-01, -4.40780058e-01,  2.60162067e-01,
        -8.05450380e-01, -9.94437403e-02,  1.80592744e+

## Choose and train model

### Subtask:
Select an appropriate machine learning model and train it on the preprocessed data.

**Reasoning**:
Import and instantiate a RandomForestClassifier model and train it using the scaled feature data `X_scaled` and the target variable `y`.

In [ ]:
from sklearn.ensemble import RandomForestClassifier

# Instantiate the model
model = RandomForestClassifier(random_state=42)

# Train the model
model.fit(X_scaled, y)

RandomForestClassifier(random_state=42)

## Evaluate model

### Subtask:
Assess the performance of the trained model using relevant metrics.

**Reasoning**:
Import the necessary metrics from sklearn.metrics for classification evaluation.

In [ ]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

**Reasoning**:
Make predictions on the training data and calculate and print the accuracy score, classification report, and confusion matrix to assess the model's performance.

In [ ]:
# Make predictions on the training data
y_pred = model.predict(X_scaled)

# Calculate and print the accuracy score
accuracy = accuracy_score(y, y_pred)
print(f'Accuracy: {accuracy:.4f}')

# Generate and print a classification report
print('Classification Report:')
print(classification_report(y, y_pred))

# Generate and print a confusion matrix
print('Confusion Matrix:')
print(confusion_matrix(y, y_pred))

Accuracy: 1.0000
Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00       212
           1       1.00      1.00      1.00       357

    accuracy                           1.00       569
   macro avg       1.00      1.00      1.00       569
weighted avg       1.00      1.00      1.00       569

Confusion Matrix:
[[212   0]
 [  0 357]]


## Make predictions (optional)

### Subtask:
Use the trained model to make predictions on new data.

**Reasoning**:
Select a few rows from the original DataFrame to represent new data, separate features and target, scale the features using the previously fitted scaler, and then use the trained model to make predictions.

In [ ]:
# Select a few rows as "new" data
new_data_df = df.sample(5, random_state=42)

# Separate features (X_new) and target (y_new)
X_new = new_data_df.drop('target', axis=1)
y_new = new_data_df['target']

# Scale the new data using the fitted scaler
X_new_scaled = scaler.transform(X_new)

# Make predictions on the scaled new data
predictions = model.predict(X_new_scaled)

# Print the predictions and compare with actual values
print("Predictions:", predictions)
print("Actual values:", y_new.values)

NameError: name 'df' is not defined

## Summary:

### Data Analysis Key Findings

* The breast cancer dataset containing 569 entries and 31 features was successfully loaded and preprocessed.
* Features were scaled using `StandardScaler` to prepare the data for model training.
* A `RandomForestClassifier` model was trained on the scaled data.
* The model achieved perfect accuracy (1.0000) on the training data, with precision, recall, and F1-score of 1.00 for both classes.
* The confusion matrix showed no misclassifications on the training data.
* Predictions made on a small sample of "new" data matched the actual values.

### Insights or Next Steps

* The perfect performance on the training data suggests potential overfitting. The next step should be to split the data into training and testing sets and evaluate the model's performance on unseen data to get a realistic assessment of its generalization ability.
* Consider hyperparameter tuning of the `RandomForestClassifier` or exploring other models to potentially improve performance and generalization on unseen data.

# Task
Split the data into training and testing sets, train a RandomForestClassifier model, evaluate the model, perform hyperparameter tuning for the RandomForestClassifier, and evaluate the tuned model.

## Split data

### Subtask:
Split the data into training and testing sets using `train_test_split`.


**Reasoning**:
Import train_test_split and split the data into training and testing sets.



In [1]:
from sklearn.model_selection import train_test_split

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

# Print the shapes of the resulting sets to verify the split
print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

NameError: name 'X_scaled' is not defined

**Reasoning**:
The previous code failed because X_scaled and y were not defined in the current session. Need to reload the data and preprocess it again to define X_scaled and y before splitting.



In [2]:
from sklearn.datasets import load_breast_cancer
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

# Load the breast cancer dataset
breast_cancer = load_breast_cancer()

# Convert to a pandas DataFrame
df = pd.DataFrame(data=breast_cancer.data, columns=breast_cancer.feature_names)
df['target'] = breast_cancer.target

# Separate features (X) and target (y)
X = df.drop('target', axis=1)
y = df['target']

# Import and instantiate StandardScaler
scaler = StandardScaler()

# Fit and transform the feature data
X_scaled = scaler.fit_transform(X)

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

# Print the shapes of the resulting sets to verify the split
print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (455, 30)
Shape of X_test: (114, 30)
Shape of y_train: (455,)
Shape of y_test: (114,)


## Train model

### Subtask:
Train the `RandomForestClassifier` model on the training data.


**Reasoning**:
Import and instantiate a RandomForestClassifier model and train it using the training data.



In [3]:
from sklearn.ensemble import RandomForestClassifier

# Instantiate the model
model = RandomForestClassifier(random_state=42)

# Train the model
model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

## Evaluate model

### Subtask:
Evaluate the trained model on the testing data using appropriate metrics.


**Reasoning**:
Make predictions on the testing data and calculate and print the accuracy score, classification report, and confusion matrix to evaluate the model's performance on unseen data.



In [4]:
# Make predictions on the testing data
y_pred_test = model.predict(X_test)

# Calculate and print the accuracy score
accuracy_test = accuracy_score(y_test, y_pred_test)
print(f'Accuracy on testing data: {accuracy_test:.4f}')

# Generate and print a classification report
print('Classification Report on testing data:')
print(classification_report(y_test, y_pred_test))

# Generate and print a confusion matrix
print('Confusion Matrix on testing data:')
print(confusion_matrix(y_test, y_pred_test))

NameError: name 'accuracy_score' is not defined

**Reasoning**:
The previous command failed because the necessary metrics functions from `sklearn.metrics` were not imported in the current session. The code block needs to import these functions before using them to evaluate the model.



In [5]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Make predictions on the testing data
y_pred_test = model.predict(X_test)

# Calculate and print the accuracy score
accuracy_test = accuracy_score(y_test, y_pred_test)
print(f'Accuracy on testing data: {accuracy_test:.4f}')

# Generate and print a classification report
print('Classification Report on testing data:')
print(classification_report(y_test, y_pred_test))

# Generate and print a confusion matrix
print('Confusion Matrix on testing data:')
print(confusion_matrix(y_test, y_pred_test))

Accuracy on testing data: 0.9649
Classification Report on testing data:
              precision    recall  f1-score   support

           0       0.98      0.93      0.95        43
           1       0.96      0.99      0.97        71

    accuracy                           0.96       114
   macro avg       0.97      0.96      0.96       114
weighted avg       0.97      0.96      0.96       114

Confusion Matrix on testing data:
[[40  3]
 [ 1 70]]


## Hyperparameter tuning

### Subtask:
Use techniques like GridSearchCV or RandomizedSearchCV to find the best hyperparameters for the model.


**Reasoning**:
Import GridSearchCV, define the hyperparameter grid, instantiate GridSearchCV with the model and grid, and fit it to the training data to find the best hyperparameters and their corresponding score.



In [6]:
from sklearn.model_selection import GridSearchCV

# Define the hyperparameter grid
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

# Instantiate GridSearchCV
grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=5, scoring='accuracy', n_jobs=-1)

# Fit GridSearchCV to the training data
grid_search.fit(X_train, y_train)

# Print the best hyperparameters and best score
print("Best hyperparameters:", grid_search.best_params_)
print("Best cross-validation accuracy:", grid_search.best_score_)

Best hyperparameters: {'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 200}
Best cross-validation accuracy: 0.9626373626373625


## Evaluate tuned model

### Subtask:
Evaluate the model with the best hyperparameters on the testing data.


**Reasoning**:
Get the best estimator from the grid search and evaluate it on the test data by calculating accuracy, classification report, and confusion matrix.



In [7]:
# Get the best model from the grid search
best_model = grid_search.best_estimator_

# Make predictions on the testing data with the best model
y_pred_tuned = best_model.predict(X_test)

# Calculate and print the accuracy score of the tuned model
accuracy_tuned = accuracy_score(y_test, y_pred_tuned)
print(f'Accuracy of tuned model on testing data: {accuracy_tuned:.4f}')

# Generate and print a classification report for the tuned model
print('Classification Report of tuned model on testing data:')
print(classification_report(y_test, y_pred_tuned))

# Generate and print a confusion matrix for the tuned model
print('Confusion Matrix of tuned model on testing data:')
print(confusion_matrix(y_test, y_pred_tuned))

Accuracy of tuned model on testing data: 0.9649
Classification Report of tuned model on testing data:
              precision    recall  f1-score   support

           0       0.98      0.93      0.95        43
           1       0.96      0.99      0.97        71

    accuracy                           0.96       114
   macro avg       0.97      0.96      0.96       114
weighted avg       0.97      0.96      0.96       114

Confusion Matrix of tuned model on testing data:
[[40  3]
 [ 1 70]]


## Summary:

### Data Analysis Key Findings

*   The data was successfully split into training and testing sets with a test size of 20%.
*   An initial `RandomForestClassifier` model achieved an accuracy of 0.9649 on the testing data.
*   Hyperparameter tuning using `GridSearchCV` identified the best parameters as `{'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 200}`, yielding a best cross-validation accuracy of approximately 0.9626.
*   The tuned `RandomForestClassifier` model achieved an accuracy of 0.9649 on the testing data, the same as the initial model.
*   The tuned model showed strong performance across other metrics on the testing data, including high precision, recall, and F1-scores for both classes, with a confusion matrix showing only 3 false positives and 1 false negative.

### Insights or Next Steps

*   Since the accuracy on the testing set did not improve after hyperparameter tuning, the default parameters or a wider range of hyperparameters could be explored.
*   Consider evaluating the model using other metrics relevant to the specific problem domain, such as ROC AUC or PR AUC, especially if class imbalance is a concern.


## Evaluate model with additional metrics

### Subtask:
Evaluate the tuned model on the testing data using ROC AUC and PR AUC.

**Reasoning**:
Import `roc_auc_score` and `precision_recall_curve` from `sklearn.metrics`, calculate the ROC AUC score, and generate the precision-recall curve data to evaluate the tuned model's performance using additional relevant metrics.

In [8]:
from sklearn.metrics import roc_auc_score, precision_recall_curve, auc

# Calculate ROC AUC
y_prob_tuned = best_model.predict_proba(X_test)[:, 1]
roc_auc = roc_auc_score(y_test, y_prob_tuned)
print(f'ROC AUC of tuned model on testing data: {roc_auc:.4f}')

# Calculate Precision-Recall AUC
precision, recall, _ = precision_recall_curve(y_test, y_prob_tuned)
pr_auc = auc(recall, precision)
print(f'PR AUC of tuned model on testing data: {pr_auc:.4f}')

ROC AUC of tuned model on testing data: 0.9959
PR AUC of tuned model on testing data: 0.9975


## Summary:

### Data Analysis Key Findings

* The data was successfully split into training and testing sets with a test size of 20%.
* An initial `RandomForestClassifier` model achieved an accuracy of 0.9649 on the testing data.
* Hyperparameter tuning using `GridSearchCV` identified the best parameters as `{'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 200}`, yielding a best cross-validation accuracy of approximately 0.9626.
* The tuned `RandomForestClassifier` model achieved an accuracy of 0.9649 on the testing data, the same as the initial model.
* The tuned model showed strong performance across other metrics on the testing data, including high precision, recall, and F1-scores for both classes, with a confusion matrix showing only 3 false positives and 1 false negative.
* The tuned model also achieved high ROC AUC (0.9959) and PR AUC (0.9975) scores on the testing data, indicating excellent discrimination ability.

### Insights or Next Steps

* Since the accuracy on the testing set did not improve after hyperparameter tuning, the default parameters or a wider range of hyperparameters could be explored. However, given the already high performance on the test set and the additional metrics, significant improvements might be difficult to achieve with this dataset and model.
* Consider evaluating the model using other metrics relevant to the specific problem domain, such as ROC AUC or PR AUC, especially if class imbalance is a concern. (Already done in the previous step).
* Further analysis of the false positives and false negatives from the confusion matrix could provide insights into specific types of misclassifications and potential areas for model improvement or data collection.
* The high performance on the testing set suggests that the model generalizes well to unseen data, despite the initial perfect score on the training data.

## Save the trained model

### Subtask:
Save the trained `best_model` using `joblib`.

**Reasoning**:
Import the `joblib` library and save the `best_model` object to a file named 'best_random_forest_model.joblib'.

In [9]:
import joblib

# Save the trained model
joblib.dump(best_model, 'best_random_forest_model.joblib')

print("Trained model saved as 'best_random_forest_model.joblib'")

Trained model saved as 'best_random_forest_model.joblib'


## Load the trained model and make predictions

### Subtask:
Load the saved model and use it to make predictions on new data.

**Reasoning**:
Import the `joblib` library and load the saved model from the 'best_random_forest_model.joblib' file. Then, use the loaded model to make predictions on a small sample of new data, remembering to scale the new data using the same scaler fitted during preprocessing.

In [10]:
import joblib
import pandas as pd

# Load the trained model
loaded_model = joblib.load('best_random_forest_model.joblib')

print("Trained model loaded successfully.")

# To make predictions on new data, you would follow these steps:
# 1. Load or create your new data in a pandas DataFrame.
# 2. Ensure the new data has the same features as the training data.
# 3. Scale the new data using the *same* scaler that was fitted to the training data.
#    (We need to ensure the 'scaler' object is available in this session or saved and loaded as well)
# 4. Use the loaded model to make predictions on the scaled new data.

# Example of making predictions on a few sample rows from the original data (as if it were new data)
# We'll reuse the 'df' DataFrame from earlier and the 'scaler' object.
# In a real deployment, you would load your new data and the saved scaler.

# Select a few rows as "new" data (assuming 'df' and 'scaler' are available from previous steps)
if 'df' in locals() and 'scaler' in locals():
    new_data_df = df.sample(5, random_state=99) # Using a different random state for variety

    # Separate features (X_new)
    X_new = new_data_df.drop('target', axis=1)

    # Scale the new data using the fitted scaler
    X_new_scaled = scaler.transform(X_new)

    # Make predictions on the scaled new data
    predictions = loaded_model.predict(X_new_scaled)

    # Print the predictions
    print("\nPredictions on sample new data:", predictions)
    # If you had the actual values for this new data, you could compare them:
    # print("Actual values:", new_data_df['target'].values)
else:
    print("\nCould not make predictions on sample data because 'df' or 'scaler' is not defined in this session.")
    print("In a real deployment, you would load your new data and the saved scaler.")

Trained model loaded successfully.

Predictions on sample new data: [1 1 1 1 1]
